crates/sui-types/src/temporary_store.rs

```diff
@@ -508,18 +508,20 @@ impl<S> TemporaryStore<S> {
         gas_object: &mut Object,
     ) -> Result<(), ExecutionError> {
         let mut objects_to_update = vec![];
-        // Also charge gas for mutating the gas object in advance.
-        let gas_object_size = gas_object.object_size_for_gas_metering();
-        gas_object.storage_rebate = gas_status.charge_storage_mutation(
-            gas_object_size,
-            gas_object_size,
-            gas_object.storage_rebate.into(),
-        )?;
-        objects_to_update.push((
-            SingleTxContext::gas(sender),
-            gas_object.clone(),
-            WriteKind::Mutate,
-        ));
+        // If the gas coin was not yet written, charge gas for mutating the gas object in advance.
+        if !self.written.contains_key(&gas_object.id()) {
```

Comment:
@dariorussi @oxade for visibility.
This was a bug just for programmable transactions, since only in that case will it already be written.

Reply:
iirc now we have vector of gas coins already, do we plan to replace `gas_object` with `gas_objects` everywhere?

Reply:
Not quite. We assume at this point that gas has already been smashed, so we only have 1 gas object to deal with. 

The flow relevant to this piece of code:
- Smash gas
- Execute the transaction
- Charge for gas